en/docs/extended_topics/user_op.md

```diff
@@ -1,414 +1,816 @@
-    AddOp(relu_grad_op);  /* step 4. 将子图中新创建的op添加到计算图中 */
-  }
-});
+void fn(BackwardOpConfContext* ctx);
+```
+
+`BackwardOpConfContext* ctx` has all information needed to generate the op.
+
+### BackwardOpConfContext
+
+The common methods and their purpose used in `BackwardOpConfContext` as follows:
+
+* `UserOpWrapper& FwOp();`: Get forward op.
+
+* `GetOp(op_name)`: Create and get the corresponding `op` based on `op_name`. `GetOp` uses a lazy init mechanism and the corresponding op is not actually created until `GetOp` is called.
+
+* `void DefineOp(op_name, fn)`：Define `fn` function of the op named `op_name`. When `ctx->GetOp(op_name)` is called, `fn` is triggered in the OneFlow for Op creation and if the op has already been created. Then the result is retrieved directly. The `fn` receives a `BackwardOpBuilder` parameter for constructing the reverse op. We will introduce `BackwardOpBuilder` later on.
+
+
+### Detailed Introduction of BackwardOpBuilder
+
+`BackwardOpBuilder` is used to build a reverse op. The fragment of above code is an example:
+
+```cpp
+ctx->DefineOp(op1_name, 
+  [&ctx](user_op::BackwardOpBuilder& builder) {
+    return builder.OpTypeName("multiply")
+        .InputBind("x", ctx->FwOp().input("in", 0)) //multiply.x <- myop.in
+        .InputBind("y", ctx->FwOp().output_grad("out", 0)) //multiply.y <- myop.out的梯度
+        .Output("out")
+        .Build();
+  });
+```
+
+In this function, we calling `Build` to build a reverse op for computing `x*dy`.
+The purpose of each operator is as follows:
+
+* `OpTypeName("multiply")` specifies the `op_type_name` of an op that is used to help us compute the reverse gradient.
+
+* `InputBind(arg_name, blob)` binds the input `arg_name` of `multiply` to the specified `blob` and can be called for multiple times. If the `arg_name` corresponds to multiple  blob which means the order of `Input` is the order of the corresponding index.
+
+* `Output(arg_name, num)` Specifies the number of output blobes that actually correspond to the `arg_name` which defaults to 1 if `num` is not filled in.
+
+* `Attr(attr_name, val) ` sets the value of the attribute which same in the registration.
+
+* After completing settings in `Build()`. Then complete the construction of the reverse op by calling `Build`.
+
+
+### Detailed Introduction of UserOpWrapper
+
+Calling `ctx->FwOp()` will return the `UserOpWrapper`of `myop` and complete the gradient binding by calling the `UserOpWrapper`.
+
+```cpp
+ctx->FwOp().InputGradBind(user_op::OpArg("in", 0), 
+  [&ctx, &op2_name]() -> const std::string& {
+    return ctx->GetOp(op2_name)
+          .output("out", 0);
+  });
+```
+
+Common methods for `UserOpWrapper` are:
+
+* `InputGradBind(input, grad_fn)`：Bind the input of the forward op and get the gradient function `grad_fn`.  OneFlow automatically determines whether `input` needs to generate a backward gradient and if needed, triggers `grad_fn` and binds the input.
+
```

Comment:
OneFlow automatically determines whether `input` needs to generate a backward gradient, if needed, it will trigger xxx